File: ai_server/ai/vct_search.py
"""VCT (Victory by Continuous Threats) search for Gomoku.

Extends VCF by also considering three-threats (moves that create open
threes). Where VCF only searches sequences of fours (branching ~1),
VCT searches mixed four+three sequences (branching ~2-3).

Key win conditions VCT catches that VCF misses:
  - Double-three: a single move creates 2+ open threes → unstoppable
  - Four-three: a move creates a four + an open three → unstoppable
  - Three-sequence: repeated three-threats that eventually reach a VCF

Algorithm:
  1. Check VCF first (faster, always higher priority).
  2. Try moves that create open threes (3 consecutive, both ends open).
  3. If a move creates double-three or four+three → instant win.
  4. Otherwise, enumerate defender's responses (the two extension cells
     where the three could become a four).
  5. For EACH defense, recursively check if attacker still has VCT.
  6. If ALL defenses lead to attacker win → this is a VCT winning move.

Limitation: only detects consecutive patterns. Gapped patterns (like
X X X _ X split-fours) require separate handling.
"""

from ai.game_logic import BOARD_SIZE, EMPTY, BLACK, WHITE, DIRECTIONS
from ai.pattern_eval import _count_consecutive
from ai.vcf_search import find_vcf, _four_info, _makes_five, _candidate_moves


def find_vct(board, attacker, max_depth=6, max_branch=6):
    """Search for a forced win via continuous threats (fours + threes).

    Returns the first winning move (r, c) or None.
    Board is mutated during search but restored before return.
    """
    # VCF is always tried first — faster and catches pure-four sequences
    vcf = find_vcf(board, attacker, max_depth=max_depth)
    if vcf is not None:
        return vcf
    return _vct_recurse(board, attacker, max_depth, max_branch)


def _get_open_threes(board, r, c, player):
    """After placing player at (r,c), return open threes created.

    Returns list of ((dr,dc), [ext_cell_1, ext_cell_2]) for each
    direction that forms an open three (3 consecutive, both ends open).
    Extension cells are where the attacker would play next to make a four.
    """
    result = []
    for dr, dc in DIRECTIONS:
        pos_c, pos_o = _count_consecutive(board, r, c, dr, dc, player)
        neg_c, neg_o = _count_consecutive(board, r, c, -dr, -dc, player)
        total = 1 + pos_c + neg_c
        open_ends = pos_o + neg_o
        if total == 3 and open_ends == 2:
            ext = []
            if pos_o:
                ext.append((r + (pos_c + 1) * dr, c + (pos_c + 1) * dc))
            if neg_o:
                ext.append((r - (neg_c + 1) * dr, c - (neg_c + 1) * dc))
            result.append(((dr, dc), ext))
    return result


def _defender_has_five(board, defender):
    """Can defender win immediately by placing one stone?"""
    for r, c in _candidate_moves(board, radius=1):
        if board[r][c] == EMPTY and _makes_five(board, r, c, defender):
            return True
    return False


def _vct_recurse(board, attacker, depth, max_branch):
    if depth <= 0:
        return None

    defender = WHITE if attacker == BLACK else BLACK

    # VCF check at every level — fours always dominate threes
    vcf = find_vcf(board, attacker, max_depth=depth)
    if vcf is not None:
        return vcf

    # Find moves that create open threes
    threat_moves = []
    for r, c in _candidate_moves(board, radius=2):
        if board[r][c] != EMPTY:
            continue
        board[r][c] = attacker
        threes = _get_open_threes(board, r, c, attacker)
        kind, _ = _four_info(board, r, c, attacker)
        board[r][c] = EMPTY

        if not threes:
            continue

        # Priority: double-three > four+three > single three
        has_four = kind in ('open_four', 'half_four')
        n_threes = len(threes)
        priority = n_threes * 10 + (5 if has_four else 0)
        threat_moves.append((priority, r, c, threes, has_four))

    threat_moves.sort(key=lambda x: -x[0])
    threat_moves = threat_moves[:max_branch]

    for _, r, c, threes, has_four in threat_moves:
        board[r][c] = attacker

        # Counter-check: can defender win immediately (five)?
        if _defender_has_five(board, defender):
            board[r][c] = EMPTY
            continue

        # Double-three or four+three: instant win
        # (defender can block at most one threat per turn)
        if len(threes) >= 2 or (has_four and len(threes) >= 1):
            board[r][c] = EMPTY
            return (r, c)

        # Single open three: enumerate defense moves
        # Defense = the extension cells of the open three
        # (if defender doesn't block one, attacker extends to open four → win)
        defense_cells = set()
        for (dr, dc), ext_cells in threes:
            for cell in ext_cells:
                if (0 <= cell[0] < BOARD_SIZE and 0 <= cell[1] < BOARD_SIZE
                        and board[cell[0]][cell[1]] == EMPTY):
                    defense_cells.add(cell)

        if not defense_cells:
            board[r][c] = EMPTY
            continue

        # For ALL defense moves, can attacker still win?
        # sorted() for deterministic iteration (matches Swift port).
        all_win = True
        for def_r, def_c in sorted(defense_cells):
            board[def_r][def_c] = defender
            sub = _vct_recurse(board, attacker, depth - 1, max_branch)
            board[def_r][def_c] = EMPTY
            if sub is None:
                all_win = False
                break

        board[r][c] = EMPTY
        if all_win:
            return (r, c)

    return None

File: ios_plugin/Sources/VctSearch.swift
import Foundation

// Port of ai_server/ai/vct_search.py. Extends VCF by also considering
// open-three threats: a move that creates two open threes, or one four
// and one open three, is unstoppable and counts as a win. Mixed four+
// three sequences catch positions VCF misses.
//
// VCF is always tried first inside the recursion — it's strictly faster
// and any VCF line is also a VCT line.

public enum VctSearch {

	/// After placing `player` at (row, col), enumerate directions that
	/// formed an open three (3 consecutive, both ends open). For each,
	/// return the extension cells the attacker would play next to make
	/// a four — these double as the cells the defender must block.
	public static func openThrees(
		board: [Int8], row: Int, col: Int, player: Int8
	) -> [(dir: (Int, Int), extensions: [(row: Int, col: Int)])] {
		var result: [(dir: (Int, Int), extensions: [(row: Int, col: Int)])] = []
		for (dr, dc) in GameLogic.directions {
			let pos = PatternEval.countConsecutive(
				board: board, row: row, col: col,
				dRow: dr, dCol: dc, player: player)
			let neg = PatternEval.countConsecutive(
				board: board, row: row, col: col,
				dRow: -dr, dCol: -dc, player: player)
			let total = 1 + pos.count + neg.count
			let openEnds = pos.isOpen + neg.isOpen
			if total == 3 && openEnds == 2 {
				var ext: [(row: Int, col: Int)] = []
				if pos.isOpen == 1 {
					ext.append((row + (pos.count + 1) * dr,
					            col + (pos.count + 1) * dc))
				}
				if neg.isOpen == 1 {
					ext.append((row - (neg.count + 1) * dr,
					            col - (neg.count + 1) * dc))
				}
				result.append(((dr, dc), ext))
			}
		}
		return result
	}

	private static func defenderHasFive(
		board: [Int8], defender: Int8
	) -> Bool {
		for (r, c) in VcfSearch.candidateMoves(board: board, radius: 1) {
			if board[GameLogic.idx(r, c)] == 0
					&& VcfSearch.makesFive(
						board: board, row: r, col: c, player: defender) {
				return true
			}
		}
		return false
	}

	/// Search for a forced win using fours + threes. Tries VCF first at
	/// every level. Returns (row, col) or nil; mutates+restores `board`.
	public static func findVct(
		board: inout [Int8], attacker: Int8,
		maxDepth: Int = 6, maxBranch: Int = 6
	) -> (row: Int, col: Int)? {
		if let win = VcfSearch.findVcf(
				board: &board, attacker: attacker, maxDepth: maxDepth) {
			return win
		}
		return vctRecurse(board: &board, attacker: attacker,
		                  depth: maxDepth, maxBranch: maxBranch)
	}

	private static func vctRecurse(
		board: inout [Int8], attacker: Int8, depth: Int, maxBranch: Int
	) -> (row: Int, col: Int)? {
		if depth <= 0 { return nil }
		let defender: Int8 = attacker == 1 ? 2 : 1

		if let win = VcfSearch.findVcf(
				board: &board, attacker: attacker, maxDepth: depth) {
			return win
		}

		// Collect moves that create at least one open three, tagged with
		// whether they also create a four.
		var threatMoves: [(priority: Int, row: Int, col: Int,
		                   threes: [(dir: (Int, Int), extensions: [(row: Int, col: Int)])],
		                   hasFour: Bool)] = []
		for (r, c) in VcfSearch.candidateMoves(board: board, radius: 2) {
			if board[GameLogic.idx(r, c)] != 0 { continue }
			board[GameLogic.idx(r, c)] = attacker
			let threes = openThrees(
				board: board, row: r, col: c, player: attacker)
			let fourInfo = VcfSearch.fourInfo(
				board: board, row: r, col: c, player: attacker)
			board[GameLogic.idx(r, c)] = 0

			if threes.isEmpty { continue }

			let hasFour = fourInfo.kind == .openFour || fourInfo.kind == .halfFour
			let priority = threes.count * 10 + (hasFour ? 5 : 0)
			threatMoves.append((priority, r, c, threes, hasFour))
		}

		threatMoves.sort { $0.priority > $1.priority }
		if threatMoves.count > maxBranch {
			threatMoves = Array(threatMoves.prefix(maxBranch))
		}

		for (_, r, c, threes, hasFour) in threatMoves {
			board[GameLogic.idx(r, c)] = attacker

			if defenderHasFive(board: board, defender: defender) {
				board[GameLogic.idx(r, c)] = 0
				continue
			}

			// Double-three or four+three: defender can only block one.
			if threes.count >= 2 || (hasFour && !threes.isEmpty) {
				board[GameLogic.idx(r, c)] = 0
				return (r, c)
			}

			// Single open three — defender must block one of the
			// extension cells; if every block still loses, we win.
			var defenseCells = Set<Int>()
			for (_, exts) in threes {
				for cell in exts {
					if cell.row >= 0 && cell.row < GameLogic.boardSize
							&& cell.col >= 0 && cell.col < GameLogic.boardSize
							&& board[GameLogic.idx(cell.row, cell.col)] == 0 {
						defenseCells.insert(cell.row * GameLogic.boardSize + cell.col)
					}
				}
			}

			if defenseCells.isEmpty {
				board[GameLogic.idx(r, c)] = 0
				continue
			}

			var allWin = true
			for packed in defenseCells.sorted() {
				let defR = packed / GameLogic.boardSize
				let defC = packed % GameLogic.boardSize
				board[GameLogic.idx(defR, defC)] = defender
				let sub = vctRecurse(board: &board, attacker: attacker,
				                     depth: depth - 1, maxBranch: maxBranch)
				board[GameLogic.idx(defR, defC)] = 0
				if sub == nil {
					allWin = false
					break
				}
			}

			board[GameLogic.idx(r, c)] = 0
			if allWin {
				return (r, c)
			}
		}

		return nil
	}
}
